File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/forge/ForgeVersionList.java
package org.wdt.wdtc.core.download.forge;


import com.google.gson.JsonArray;
import com.google.gson.annotations.SerializedName;
import lombok.Getter;
import org.wdt.utils.gson.JsonArrayUtils;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.download.infterface.VersionListInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.utils.URLUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ForgeVersionList implements VersionListInterface {
  private final Launcher launcher;


  public ForgeVersionList(Launcher launcher) {
    this.launcher = launcher;
  }

  public String getForgeListUrl() {
    return "https://bmclapi2.bangbang93.com/forge/minecraft/" + launcher.getVersionNumber();
  }

  @Override
  public List<VersionJsonObjectInterface> getVersionList() throws IOException {
    List<VersionJsonObjectInterface> VersionName = new ArrayList<>();
    JsonArray VersionList = JsonArrayUtils.parseJsonArray(URLUtils.getURLToString(getForgeListUrl()));
    for (int i = 0; i < VersionList.size(); i++) {
      VersionName.add(JsonObjectUtils.parseObject(VersionList.get(i).getAsJsonObject(), ForgeVersionJsonObjectImpl.class));
    }
    return VersionName;
  }

  @Getter
  public static class ForgeVersionJsonObjectImpl implements VersionJsonObjectInterface {
    @SerializedName("version")
    private String versionNumber;
    @SerializedName("modified")
    private String modified;
    @SerializedName("mcversion")
    private String mcversion;

    @Override
    public String getVersionNumber() {
      return versionNumber;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      ForgeVersionJsonObjectImpl that = (ForgeVersionJsonObjectImpl) o;
      return Objects.equals(versionNumber, that.versionNumber) && Objects.equals(mcversion, that.mcversion);
    }

    @Override
    public int hashCode() {
      return Objects.hash(versionNumber, mcversion);
    }
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/forge/ForgeVersionList.kt
package org.wdt.wdtc.core.download.forge

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.parseJsonArray
import org.wdt.utils.gson.parseObject
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.download.infterface.VersionListInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.utils.toURL
import java.net.URL
import java.util.*

class ForgeVersionList(private val version: Version) : VersionListInterface {
  private val forgeListUrl: URL
    get() = "https://bmclapi2.bangbang93.com/forge/minecraft/${version.versionNumber}".toURL()
	
	override val versionList: LinkedList<VersionsJsonObjectInterface>
		get() = LinkedList<VersionsJsonObjectInterface>().apply {
      forgeListUrl.toStrings().parseJsonArray().forEach {
        add(it.asJsonObject.parseObject<ForgeVersionsJsonObjectImpl>())
      }
    }


  class ForgeVersionsJsonObjectImpl(
    @field:SerializedName("version")
    override val versionNumber: String,

    @field:SerializedName("modified")
    val modified: Date,
  ) : VersionsJsonObjectInterface {

    @SerializedName("mcversion")
    val mcversion: String? = null
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other == null || javaClass != other.javaClass) return false
      val that = other as ForgeVersionsJsonObjectImpl
      return versionNumber == that.versionNumber && mcversion == that.mcversion
    }

    override fun hashCode(): Int {
      return Objects.hash(versionNumber, mcversion)
    }
  }
}
